docs: more content around transactions, retries and merge behaviour (#1065)
neo4j-contrib/neo4j-spark-connector: docs/modules/ROOT/pages/faq.adoc

In general, the solution is to repartition the DataFrame before writing it to Neo4j, to avoid
multiple partitioned writes from locking the same nodes and relationships.

Deadlocks are retryable, so raising `transaction.retries` and setting a non-zero `transaction.retry.timeout` absorbs occasional contention at the cost of throughput.
See xref:write/transactions.adoc#merge-safety[] for how to partition your data so the contention does not arise in the first place.

=== My write job failed halfway through. What state is the database in?

It holds everything that was committed before the failure.

The connector commits one transaction per batch of `batch.size` rows per partition, with no enclosing transaction, so a failed job leaves an arbitrary prefix of each partition's rows in the database.
Only the batch that was in flight is rolled back.

> Do not rerun a failed write as if nothing had happened, unless the write is idempotent.

Reviewer: ```suggestion
Do *not* rerun a failed write unless the write is idempotent.
```

See xref:write/transactions.adoc#partial-failures[].

=== My write used MERGE and succeeded, but the graph has duplicate nodes. Why?

> Almost always because there is no uniqueness or key constraint on the properties being merged on.

Reviewer 2: ```suggestion
This happens almost always because there is no uniqueness or key constraint on the properties being merged on.
```

`MERGE` matches and then creates, and that pair is only atomic if a constraint exists on the merged properties.
When several Spark partitions merge the same key at the same time, they can all fail to match and all create, which produces duplicates and no error.

Create the constraint before writing, with the xref:write/schema-optimization.adoc[schema optimization options] or a xref:write/query.adoc#script-option[`script` option].
See xref:write/transactions.adoc#merge-safety[].

=== Can I use `CALL { … } IN TRANSACTIONS` in the `query` option to batch my own writes?

> No. The connector runs your query inside an explicit transaction, and Neo4j does not permit that clause in an explicit transaction, so the write fails.

Reviewer: ```suggestion
No.
The connector runs the query inside an explicit transaction, and Neo4j does not allow that clause in an explicit transaction, so the write fails.
```

You can read more xref:read/schema.adoc[here].

> === I am getting an error _TableProvider implementation org.neo4j.spark.DataSource cannot be written with ErrorIfExists mode, please use Append or Overwrite modes instead_. What should I do?

Reviewer 3: Can't get why we removing this ?)

Author: Reverted.
